### PR TITLE
[Messenger] Move topics resolution to a dedicated service

### PR DESCRIPTION
**Pull request by skalpa**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | small
| Deprecations? | no
| Tests pass?   | not yet
| Fixed tickets | --
| License       | MIT
| Doc PR        | --

This goes on top of @nicolas-grekas recent refactoring of the handlers locators.

It's obviously incomplete (the bundle changes and tests aren't here) but I wanted to get feedback as early as possible and see if it was worth finishing. 

Disclaimer: I know that my choice to use the word "topics" will be controversial, but I couldn't find anything better. Suggestions welcome ;).

- Implementations of `TopicsResolverInterface` extract _topics_ from messages. Topics are just strings with no meaning for the system, that are mapped to handlers and senders by `HandlersLocator` and `SendersLocator`. Ie: the keys of the array returned by `MessageSubscriberInterface::getHandledMessages()` are topics.
- `ClassTopicsResolver` returns the list of classes and interfaces implemented by the message class. It mimics/replaces the static `HandlersLocator::listTypes()`.
- `NameTopicsResolver` extracts a message name from the message. The name is retrieved from an attached `NameStamp` if present, or defaults to the message class. (I don't have a use for that, but there seems to be some interest so I just added it).
- Both `HandlersLocator` and `SendersLocator` now delegate the topics resolution, then associate the topics returned by their `TopicsResolverInterface` to handlers or senders.

Imho that offers a better extension point than the locator interfaces, and helps prevent code duplication between the senders and handlers locators (see the static internal method I removed). Personally, I would even be for removing `HandlersLocatorInterface` and `SendersLocatorInterface`, and marking the current implementations as `@internal`, as I couldn't find a use-case for them anymore.

The PR is not ready to be reviewed, I'll finish if we can reach a consensus.

Status: Needs work

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I'm not sure we need this. You can decorate a SendersLocator + yield additional senders.

**Comment by skalpa** [comment]
It's really more complex with the current implementation.

The point is that locators actually have two responsibilities:
- They extract topics (currently: classes and interfaces names) from messages. This is shared between both locators, and what I (and I'm convinced most people) will want to plug into.
- They map this list of topics to collections of handlers or senders (depending on the type of locator)

If topics resolution isn't behind an interface people can implement, they will have to decorate all the services that depend on it.

Ie, the most simplistic content-based routing will look like something like this.

**With the PR**:
```php
class ColorTopicsResolver implements TopicsResolverInterface
{
    public function getTopics(Envelope $envelope): iterable
    {
        if ($envelope->getMessage() instanceof ColoredInterface) {
            yield $envelope->getMessage()->getColor();
        }
    }
}

class RedHandler implements MessageSubscriberInterface
{
    public function getHandledMessages(): iterable
    {
        yield Colors::RED;
    }
    public function __invoke($message): void {}
}
```

**Without the PR**:
- You'd have to decorate both `SendersLocator` and `HandlersLocator`
- At that point you'll have the choice between: ~~duplicating code~~, implementing something similar to what this PR proposes in userland, ~~using a static method~~
- If you want handlers to be able to subscribe to topics, you'll have to completely duplicate the `MessageSubscriberInterface` system and its associated code from [`MessengerPass`](https://github.com/symfony/symfony/blob/2bfcaeb31ff3b878734b0951851d25abf2b3321e/src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php#L79), as the current system is [tied to classes and interfaces](https://github.com/symfony/symfony/blob/2bfcaeb31ff3b878734b0951851d25abf2b3321e/src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php#L124).
- To be fair, the last point could be solved separately. But fixing the bundle so it becomes compatible with something that is not defined, documented or facilitated anywhere else feels weird and incomplete.
- I intend to PR a third locator at some point: I use a `TestHandlerLocator` that lets me inject handlers during runtime an "mock" message handling, so I can test complex scenarii where my handlers dispatch more messages. Such a locator would be able to use the users' custom topics resolvers transparently. Without the new interface, users who need to customize the resolution process will _also_ have to decorate it to use it.

TL;DR Two responsibilities, two interfaces, two types of services, and a clear, documented and extension-friendly answer to "what do message subscribers actually subscribe to ?"

**Comment by nicolas-grekas** [comment]
Why wouldn't you make your red color an interface and add it to the message ? That would achieve the same with zero extra code. Note also that getHandledMessages() *must* return an existing class or interface name.
I would suggest to resist adding more concepts (thus more complexity) before someone has an actual real world use case.

**Comment by Tobion** [comment]
So what this issue is about: being able to route messsages (to senders and handler) based on the message contents.

I can see that this might be useful because sometimes you cannot use different message classes with different interfaces as the messages are fixed. For example we get messages from a third party like

```
{
    "entity-id": "myid",
    "event-type: "insert|update|delete"
}
```

So in the processor we would do

```
switch ($message->getEventType()) {
	case 'insert':
		// fallthrough
	case 'update:
		$this->importProduct($message);
		break;
	case 'delete:
		$this->deleteProduct($message);
		break;
}
```

So with the proposal here would could have different handlers based on the "event-type" of the message. But I'm not sure it's worth it. In the end you can achieve the same with a single handler and than delegating like done with the switch.

The same applies to the example given by @skalpa. Instead of a `RedHandler` you would create a `ColorHandler` and then delegate based on the color.

So I feel adding another layer to the routing/resolvingm does not seem worth it because people can achieve the same thing in user-land.
